refactor .run to store stdout, stderr, and exitcode

petersenallen opened the pull request.

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -292,60 +293,87 @@ def inspect
   # * <command> - the command to run (sudo will be prepended if specified in object instantiation)
   # * [expected_exitcode] - allows for non-0 exit codes to be returned without requiring exception handling
-  def run(command, expected_exitcode=[0])
+  def run( command, expected_exitcode = 0, sudo = self.uses_sudo? )
```
can we update the input parameters to include `sudo`? also, the context for `<command>` is correct, we still sudo if that was in object instantiation, and that is default, but as the parameter overrides it, i think we should remove that bit

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -292,60 +293,87 @@ def inspect
+      :exitcode          => nil,
+    }
+
+    cmd[:final_command] = cmd[:sudo] ? sprintf( 'sudo bash -c "%s"', command ) : command
```
any reason not to do this in the hash definition itself? think this is a holdover from before we made `sudo` a parameter

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -292,60 +293,87 @@ def inspect
-    else
-      @exitcode = 1
+    if cmd[:stdout].nil?
+      cmd[:stdout]   = "error gathering output, last logged output[#{self.get_ssh_stdout()}]"
```
not sure it makes sense to include `self.get_ssh_stdout()` if the block here will only trigger if it is `nil`.. maybe change this to `.get_ssh_stderr()`? although setting that to be `:stdout` might be more confusing.

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -292,60 +293,87 @@ def inspect
+    self.ssh_stdout.push(   cmd[:stdout]   )
+    self.ssh_stderr.push(   cmd[:stderr]   )
+    self.ssh_exitcode.push( cmd[:exitcode] )
+    @logger.debug( sprintf( 'output: [%s]',     cmd[:stdout] ) )
```
can we change 'output' here to 'ssh_stdout'?

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -292,60 +293,87 @@ def inspect
+    unless cmd[:expected_exitcode].member?( cmd[:exitcode] )
       # TODO technically this could be a 'LocalPassthroughExecutionError' now too if local passthrough.. should we update?
-      raise RemoteExecutionError.new("output[#{output}], exitcode[#{@exitcode}], expected[#{expected_exitcode}]")
+      raise RemoteExecutionError.new("output[#{cmd[:stdout]}], exitcode[#{cmd[:exitcode]}], expected[#{cmd[:expected_exitcode]}]")
```
should we include `cmd[:stderr]` here as well? if something went wrong, and the program being called is following The Rules, it's likely that STDERR will have some useful information

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -292,60 +293,87 @@ def inspect
+        channel.on_data do |ch, data|
+          cmd[:stdout] << data
+        end
+        channel.on_extended_data do |ch, type, data|
```
can you change `ch` to `_ch` and `type` to `_type` here to make it clear that we're not using these parameters?

also, how do you feel about rewriting these 3 blocks as:

```rb

channel.on_data { |_ch, data| cmd[:stdout] << data }
channel.on_extended_data { |_ch, _type, data| cmd[:stderr] << data }
channel.on_request('exit-status') do { |_ch, data| cmd[:exitcode] = data.read_long } 
```

i'm not sure i like it better based on indentation, but single line do blocks are something i've been moving away from

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -704,7 +732,7 @@ def restart(wait=nil, expected_exitcodes = [0])
   #
   # returns STDOUT|STDERR, raises Rouster::LocalExecutionError on non 0 exit code
-  # sets @exitcode
+  # pushes to @ssh_exitcode
```
i know this is legacy, but now it feels even weirder - this is not an SSH exit code, it's a local run exit code. given the use cases for this method (all local, and mostly related to environment sanity), how do you feel about just not returning the exitcode - if it isn't 0, we're throwing anyway, and if it is, well, that's just expected.

chorankates-sfdc commented on lib/rouster.rb:
```diff
@@ -718,14 +746,14 @@ def _run(command)
     File.delete(tmp_file) or raise InternalError.new(sprintf('unable to delete [%s]: %s', tmp_file, $!))
 
-    self.output.push(output)
+    self.ssh_stdout.push(output)
```
think we need to take a different tact here:
  * create a different class variable to hold output from non-SSH runs OR
  * don't cache output of exitcodes/output for non-SSH runs and just return it from this method

in this case we can't/don't distinguish between STDERR and STDOUT, so i'd be fine if the class variable didn't either

chorankates-sfdc commented on test/unit/puppet/test_get_puppet_star.rb:
```diff
@@ -20,7 +20,7 @@ def setup
 
     output = "[1;35merr: Could not retrieve catalog from remote server: Error 400 on SERVER: no result for [api_vip] in Hiera (YAML/CMS)[0m\nfizzybang\n[1;35merr: Could not retrieve catalog from remote server: Error 400 on SERVER: no result for [foobar] in Hiera (YAML/CMS)[0m"
-    @app.output.push(output)
+    @app.ssh_stderr.push(output)
```
more curiosity than review, does that output actually go to STDERR?

petersenallen commented:
Changed made after code review:

* Updated parameter comments for .run to reflect sudo changes
* Including :final_command directly in cmd hash instantiation
* Including the both stdout & stderr in the message given when stdout.nil?
* Dropped storing stdout & exitcode when executing ._run
* Style refactor of .remote_exec
